File: Module - Services/Services1/app/src/main/java/com/javadude/services1/MainActivity2.java
package com.javadude.services1;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import androidx.appcompat.app.AppCompatActivity;
import android.view.View;

public class MainActivity2 extends AppCompatActivity {

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
	}

	@Override
	protected void onStart() {
		super.onStart();
		Intent intent = new Intent(this, BoundService.class);
		bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
	}

	@Override
	protected void onStop() {
		Intent intent = new Intent(this, BoundService.class);
		stopService(intent);
		super.onStop();
	}

	public void onResetPressed(View view) {
		if (boundService != null)
			boundService.reset();
	}

	private BoundService.BoundServiceBinder boundService;

	private ServiceConnection serviceConnection = new ServiceConnection() {
		@Override
		public void onServiceConnected(ComponentName name, IBinder service) {
			boundService = (BoundService.BoundServiceBinder) service;
		}

		@Override
		public void onServiceDisconnected(ComponentName name) {
			boundService = null;
		}
	};
}

File: Module - Services/Services1/app/src/main/java/com/javadude/services1/MainActivity2.kt
package com.javadude.services1

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import kotlinx.android.synthetic.main.activity_main.*

// Activity that starts a bound service and sends it "reset" requests
class MainActivity2 : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        reset_button.setOnClickListener {
            binder?.reset()
        }

        other_activity_button.setOnClickListener {
            startActivity(Intent(this, MainActivity2a::class.java))
        }
    }

    override fun onStart() {
        Log.d("MainActivity2", "onStart")
        super.onStart()
        bindService(Intent(this, BoundService1::class.java),
            serviceConnection, Context.BIND_AUTO_CREATE)
    }

    override fun onStop() {
        Log.d("MainActivity2", "onStop")
        unbindService(serviceConnection)
        super.onStop()
    }


    private var binder : BoundService1.BoundServiceBinder? = null

    private val serviceConnection = object : ServiceConnection {
        override fun onServiceDisconnected(p0: ComponentName?) {
            this@MainActivity2.binder = null
        }

        override fun onServiceConnected(p0: ComponentName?, binder: IBinder?) {
            this@MainActivity2.binder = binder as BoundService1.BoundServiceBinder
        }
    }
}

